hasktorch/src/Data/HList.hs

```diff
@@ -69,6 +69,22 @@ instance (Monoid a, Monoid (HList as)) => Monoid (HList (a ': as)) where
 class Apply f a b where
   apply :: f -> a -> b
 
+-- | Stronger version of `Apply` that allows for better inference of the return type
+class Apply' f a b | f a -> b where
```

Comment:
Might also be nice to add a haddock to this module that briefly describes some of the uses of this machinery in the domain (ie, "HLists" let us carry around parameters of different static size as well as "stack" different "types" of things).

Reply:
yes, this module has been growing for a while now, and it is very opaque